### ISSUE TITLE
Imputing data inquiry

### ISSUE THREAD

**Issue by amblee0306** [stefaniaebli/simplicial_neural_networks#3]
Hi there,

Firstly, thanks for your work. I have a few confusion regarding the paper and would like to clarify.
In the paper, the model seems to be tasked to impute data `where some portion of the values were replaced with a constant`. 
However, i do not see how the data is imputed. This is because the paper introduced simplicial convolution which seemed to mean learning a kind of filter. Does imputing the right values just mean filtering the input p-cochains to have the 'right' value?

Another question is that, is there a reason why we impute data on the co-chains instead of chains?

Thanks

Regards :)

**Comment by stefaniaebli** [comment]
Hi,

Thank you very much for your message and interest. 

Yes, in the task of imputing missing data the input of the simplicial neural network is a p-cochain in which the value of the missing data are replaced by a constant. Then the SNN is imputing the missing values by learning the filters. In other words, the SNN learns the weights associated to the the filters which are low-degree polynomial in the p-Laplacian (see page 3 of the paper the last and second to last equation).


For your second question, I can see how one can interpret what we are doing as imputing data on the simplices or on a basis of the chains. Since p-cochains are functions on the p-simplices we see the task as having a real valued function on the p-simplices where some data is missing, This is the reason we say we are imputing data on the p-cochains.

I am happy to answer any further questions.

Cheers,
Stefania

**Comment by amblee0306** [comment]
Hi Stefania,

It is very clear, thanks for clarifying!

**Comment by amblee0306** [comment]
Hi there

I would like to clarify a little more after having a look at the code. Please correct me if i m wrong. 
The input is the p-cochains means for instance, for the 1-cochains is the vector consisting of all the 1-cochains with their respective citation values in their respective entry position in the vector? taking the complex in figure 4 for example, the 1-cochain input to the model would be a vector [150, 100, 10 ,100 ,0,4]?

Thanks in advance

**Comment by stefaniaebli** [comment]
Hi,

Yes, it is exactly as you said. The 1-cochain is just a vector of dimension the number of edges which entries are the number of citations.

Best wishes,
Stefania

**Comment by amblee0306** [comment]
Hi Stefania,

Thanks for the quick reply, really appreciate that. 
Can I also clarify the file `150250_downsampled.npy` with you? I checked the size and it only had 71 papers but the paper mentioned utilising 352 0-simplices which implies 352 papers. Can I know how can i get the biadjacency matrix for the 352 papers (with their respective edges with author nodes) that you have utilised?
This is because i checked the `paper_author_biadjacency.npz` given by the link https://zenodo.org/record/4144319#.YJDy_X0zYSI and the size was `[241412 741665]`. So I believe that it is further downsampled for you experiments. If it was further downsampled using the papers nodes in `150250_downsampled.npy`, then should only have 71 nodes instead of 352? However, I saw after running `s2_5_bipartite_to_complex.py` the number of simplices[0] was 379, which I am not sure how did it result in that. 

```
simplices, cochains, signals_top = bipart2simpcochain(adjacency, citations, indices_x=downsample_papers, dimension=11)
print(len(list(simplices[0]))) # 379
```

Please correct me if I misunderstood something. 

Thanks

![image](https://user-images.githubusercontent.com/19768905/116978615-8a783c80-acf6-11eb-8fa9-9d97424a741a.png)

![image](https://user-images.githubusercontent.com/19768905/116978940-f5c20e80-acf6-11eb-88bf-3dae9c5ca5fd.png)

**Comment by stefaniaebli** [comment]
Hi,

Yes, the file 150250_downsampled.npy has only 71 papers. The 0-simplices represents authors and not papers so what we got is 71 papers with 352 authors that are going to represent the 0-simplices. To be more precise the 71 papers we downsampled are going to represent the maximal simpleces in the coauthorship complex.  

Yes. we downsampled the number of papers to focus only on 71 papers, the indices of the downsampled papers are contained in the file 150250_downsampled.npy. Then if you wanna look at the bipartite graph of these papers it is only necessary to take the big bipartite matrix and restrict to these indices as the code below shows

adjacency = scipy.sparse.load_npz('../data/s2_2_bipartite_graph/paper_author_biadjacency.npz')
downsample_papers=np.load('../data/s2_3_collaboration_complex/'+str(s_node)+'_downsampled.npy')
bipartite_downsampled=adjacency [downsample_papers]

Cheers,
Stefania
